server/datastore/mysql/mysql.go

```diff
@@ -314,42 +315,72 @@ func (d *Datastore) MigrateData(ctx context.Context) error {
-		return 0, errors.New("unexpected nil migrations list")
+		return nil, errors.New("unexpected nil migrations list")
 	}
-
 	lastTablesMigration, err := tables.MigrationClient.Migrations.Last()
 	if err != nil {
-		return 0, fmt.Errorf("missing tables migrations: %w", err)
+		return nil, fmt.Errorf("missing tables migrations: %w", err)
 	}
-
 	currentTablesVersion, err := tables.MigrationClient.GetDBVersion(d.writer.DB)
 	if err != nil {
-		return 0, fmt.Errorf("cannot get table migration status: %w", err)
+		return nil, fmt.Errorf("cannot get table migration status: %w", err)
 	}
-
 	lastDataMigration, err := data.MigrationClient.Migrations.Last()
 	if err != nil {
-		return 0, fmt.Errorf("missing data migrations: %w", err)
+		return nil, fmt.Errorf("missing data migrations: %w", err)
 	}
-
 	currentDataVersion, err := data.MigrationClient.GetDBVersion(d.writer.DB)
 	if err != nil {
-		return 0, fmt.Errorf("cannot get data migration status: %w", err)
+		return nil, fmt.Errorf("cannot get data migration status: %w", err)
 	}
-
+	// The following assumes there cannot be migrations missing on "table"
+	// and database being ahead on "data" (and vice-versa).
 	switch {
 	case currentDataVersion == 0 && currentTablesVersion == 0:
-		return fleet.NoMigrationsCompleted, nil
-
-	case currentTablesVersion != lastTablesMigration.Version ||
-		currentDataVersion != lastDataMigration.Version:
-		return fleet.SomeMigrationsCompleted, nil
-
+		return &fleet.MigrationStatus{
+			StatusCode:   fleet.NoMigrationsCompleted,
+			TableVersion: currentTablesVersion,
+			DataVersion:  currentDataVersion,
+			MissingTable: getMissing(tables.MigrationClient.Migrations, currentTablesVersion),
+			MissingData:  getMissing(data.MigrationClient.Migrations, currentDataVersion),
+		}, nil
+	case currentTablesVersion == lastTablesMigration.Version &&
+		currentDataVersion == lastDataMigration.Version:
+		return &fleet.MigrationStatus{
+			StatusCode:   fleet.AllMigrationsCompleted,
+			TableVersion: currentTablesVersion,
+			DataVersion:  currentDataVersion,
+		}, nil
+	case currentTablesVersion > lastTablesMigration.Version ||
+		currentDataVersion > lastDataMigration.Version:
+		return &fleet.MigrationStatus{
+			StatusCode:   fleet.DatabaseVersionAhead,
+			TableVersion: currentTablesVersion,
+			DataVersion:  currentDataVersion,
+		}, nil
 	default:
-		return fleet.AllMigrationsCompleted, nil
+		// currentTablesVersion < lastTablesMigration.Version ||
+		// currentDataVersion < lastDataMigration.Version
+		return &fleet.MigrationStatus{
+			StatusCode:   fleet.SomeMigrationsCompleted,
+			TableVersion: currentTablesVersion,
+			DataVersion:  currentDataVersion,
+			MissingTable: getMissing(tables.MigrationClient.Migrations, currentTablesVersion),
+			MissingData:  getMissing(data.MigrationClient.Migrations, currentDataVersion),
+		}, nil
+	}
+}
+
+func getMissing(migrations goose.Migrations, current int64) []int64 {
```

Comment:
If I'm understanding this correctly, this detects newer migrations that haven't been added. But it's missing the case where there's a missing migration somewhere before.

Could we make it so it loads all the migrations ran from the db status table, and then compares them to the migrations that we have recorded in `{tables,data}.MigrationClient.Migrations` and see if any are missing?

Reply:
Ah, good catch! Will take a look.

Reply:
Done. We would also need a `fleet prepare db --apply-single 12345678` option, right?

Reply:
UPDATE: It seems that out-of-order migrations could only happen if there's a human error during release (human doing the release missed to re-shuffle the migrations). In any case this new `fleetctl debug` command and `serve`/`prepare db` will at least warn about this (current `main` won't warn).